tests/src/FunctionalJavascript/ContactSubmissionTest.php

```diff
@@ -0,0 +1,196 @@
+<?php
+
+namespace Drupal\Tests\webform_civicrm\FunctionalJavascript;
+
+use Behat\Mink\Exception\ElementNotFoundException;
+
+/**
+ * Tests submitting a Webform with CiviCRM and a single contact.
+ *
+ * @group webform_civicrm
+ */
+final class ContactSubmissionTest extends WebformCivicrmTestBase {
+
+  /**
+   * {@inheritdoc}
+   *
+   * During tests configuration schema is validated. This module does not
+   * provide schema definitions for its handler.
+   *
+   * To fix: webform.webform.civicrm_webform_test:handlers.webform_civicrm.settings
+   *
+   * @see \Drupal\Core\Test\TestSetupTrait::getConfigSchemaExclusions
+   */
+  protected static $configSchemaCheckerExclusions = [
+    'webform.webform.civicrm_webform_test',
+  ];
+
+  /**
+   * Test submitting a contact.
+   *
+   * @dataProvider dataContactValues
+   */
+  public function testSubmitWebform($contact_type, array $contact_values) {
+    $this->assertArrayHasKey('contact', $contact_values, 'Test data must contain contact');
+    $this->assertArrayHasKey('first_name', $contact_values['contact'], 'Test contact data must contain first_name');
+    $this->assertArrayHasKey('last_name', $contact_values['contact'], 'Test confact data must contain last_name');
+
+    // @todo change to login as admin then logout, then login as test user for submission
+    $this->drupalLogin($this->testUser);
+    $this->drupalGet($this->webform->toUrl('settings'));
+    $this->getSession()->getPage()->clickLink('CiviCRM');
+    // @todo Randomly this fails saying that the checkbox does not exist.
+    $this->assertSession()->waitForField('Enable CiviCRM Processing');
+    $this->htmlOutput();
+    $this->getSession()->getPage()->checkField('Enable CiviCRM Processing');
+    $this->getSession()->getPage()->selectFieldOption('1_contact_type', strtolower($contact_type));
+    $this->assertSession()->assertWaitOnAjaxRequest();
+
+    // @todo this should be refactored as its duplicated for each other type.
+    if (isset($contact_values['email'])) {
+      $this->assertTrue(is_array($contact_values['email']));
+      $this->assertTrue(isset($contact_values['email'][0]));
+      $this->getSession()->getPage()->selectFieldOption('contact_1_number_of_email', count($contact_values['email'][0]));
+      $this->assertSession()->assertWaitOnAjaxRequest();
+      $this->htmlOutput();
+      $this->assertSession()->checkboxChecked('civicrm_1_contact_1_email_email');
+    }
+    if (isset($contact_values['website'])) {
+      $this->assertTrue(is_array($contact_values['website']));
+      $this->assertTrue(isset($contact_values['website'][0]));
+      $this->getSession()->getPage()->selectFieldOption('contact_1_number_of_website', count($contact_values['website'][0]));
+      $this->assertSession()->assertWaitOnAjaxRequest();
+      $this->htmlOutput();
+      $this->assertSession()->checkboxChecked('civicrm_1_contact_1_website_url');
+    }
```

Comment:
This is a lot of duplicate code, especially when Phone, Messaging, and Address are added